internal/functions/serve/watcher.go

```diff
@@ -0,0 +1,368 @@
+			cwd, CWDerr := os.Getwd()
+			if CWDerr != nil {
+				utils.Warning("could not get current working directory: %v. Using relative path for watcher: %s", CWDerr, functionsDir)
+			} else {
+				absFunctionsPath = filepath.Join(cwd, functionsDir)
+			}
+		}
+	}
+	absFunctionsPath = filepath.Clean(absFunctionsPath)
+
+	// Add all import dependencies to the watcher (including the functions directory itself)
+	if err := addImportDependenciesToWatcher(watcher, watchedDirs, fsys); err != nil {
+		utils.Warning("could not add import dependencies to watcher: %v", err)
+		// Return the watcher but an empty path to signal that watching isn't properly set up.
+		return watcher, "", nil
+	}
+
+	// If no directories are being watched (e.g., functions directory doesn't exist), return empty path
+	if len(watchedDirs) == 0 {
+		utils.Info(1, "No directories found to watch")
+		return watcher, "", nil
+	}
+
+	utils.Info(1, "File watcher initialized, watching %d directories", len(watchedDirs))
+	return watcher, absFunctionsPath, nil
+}
+
+// runFileWatcher listens for events from the watcher, debounces them, and signals for a restart.
+func runFileWatcher(ctx context.Context, watcher *fsnotify.Watcher, watchedPath string, restartChan chan<- struct{}, watchedDirs map[string]bool, fsys afero.Fs) {
+	var restartTimer *time.Timer // Timer for debouncing restarts
+
+	if watchedPath == "" {
+		return
+	}
+
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+
+			if isIgnoredFileEvent(event.Name, event.Op) {
+				utils.Debug("Ignoring file event: %s (%s)", event.Name, event.Op.String())
+				continue
+			}
+
+			// Handle file change events that should trigger a restart
+			isSignificantEventForRestart := event.Has(fsnotify.Write) || event.Has(fsnotify.Create) || event.Has(fsnotify.Remove) || event.Has(fsnotify.Rename)
+
+			if isSignificantEventForRestart {
+				fileName := filepath.Base(event.Name)
+
+				if fileName == "config.toml" {
+					utils.Info(2, "Config file change detected: %s (%s) - will reload configuration", event.Name, event.Op.String())
+				} else {
+					utils.Info(2, "File change detected: %s (%s)", event.Name, event.Op.String())
+				}
+
+				// Re-add import dependencies to catch any new dependencies
+				if err := addImportDependenciesToWatcher(watcher, watchedDirs, fsys); err != nil {
+					utils.Warning("could not update import dependencies: %v", err)
+				}
+
+				if restartTimer != nil {
+					restartTimer.Stop()
+				}
+				restartTimer = time.AfterFunc(debounceDuration, func() {
+					restartChan <- struct{}{}
+				})
+			}
+
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return
+			}
+			utils.Warning("File watcher error: %v", err)
+		case <-ctx.Done():
+			if restartTimer != nil {
+				restartTimer.Stop()
+			}
+			return
+		}
+	}
+}
+
+type FileWatcher interface {
+	Watch(ctx context.Context, fsys afero.Fs) (<-chan struct{}, <-chan error)
+	Close() error
+}
```

Comment:
This file watcher is tightly coupled with our internals of functions and config paths. I think we can go with a more general implementation

```suggestion
type FileWatcher interface {
	Watch(ctx context.Context) (<-chan struct{}, <-chan error)
	SetWatchPaths(watchPaths []string, fsys afero.Fs) error
	Close() error
}
```

The main responsibilities for file watcher are
1. Denounce fs events
2. Support recursive directories


After receiving an event from restart channel, it's the caller's responsibility to set the new watch paths.

Reply:
> if we automatically watch recursively all the directory from there we'll end up with a lot of watchers which might cause issues.

Good point. I wonder how nextjs solve it in their local dev?

Perhaps we can set a max recursion depth or other cleverer ways of pruning.